### PR TITLE
Fix out-of-bounds access for 1x1 tridiagonal matrices and add shape checks

### PR DESCRIPTION
**Pull request by srinjoy933**
solve #1161

### CODE REVIEW

File: src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp [hunk 1]
```diff
@@ -283,16 +286,14 @@ submodule (stdlib_specialmatrices) tridiagonal_matrices
         ${t1}$, intent(in) :: alpha
         type(tridiagonal_${s1}$_type), intent(in) :: A
         type(tridiagonal_${s1}$_type) :: B
-        B = tridiagonal(A%dl, A%dv, A%du)
-        B%dl = alpha*B%dl; B%dv = alpha*B%dv; B%du = alpha*B%du
+        B = tridiagonal(alpha*A%dl, alpha*A%dv, alpha*A%du)
     end function
```
[the remark this answers is not in the document]
Comment (loiseaujc) [reply src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp:270]: Copilot is right here. `tridiagonal(alpha*A%dl, alpha*A%dv, alpha*A%du)` typically will trigger three copies (one for each component) followed by the allocation of the internal arrays for `B`. In contrast,

```
B = tridiagonal(A%dl, A%dv, A%du)
B%dl = alpha*B%dl; B%dv = alpha*B%dv; B%du = alpha*B%du
```

only incurs the allocation of the internal arrays followed by in-place operations. Please revert this change.

File: src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp [hunk 2]
```diff
@@ -301,17 +302,21 @@ submodule (stdlib_specialmatrices) tridiagonal_matrices
         type(tridiagonal_${s1}$_type), intent(in) :: A
         type(tridiagonal_${s1}$_type), intent(in) :: B
         type(tridiagonal_${s1}$_type) :: C
-        C = tridiagonal(A%dl, A%dv, A%du)
-        C%dl = C%dl + B%dl; C%dv = C%dv + B%dv; C%du = C%du + B%du
+        
+        if (A%n /= B%n) error stop "ValueError: tridiagonal matrices must have the same dimension to be added"
+        
+        C = tridiagonal(A%dl + B%dl, A%dv + B%dv, A%du + B%du)
```
[the remark this answers is not in the document]
Comment (loiseaujc) [reply src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp]: `C = tridiagonal(A%dl + B%dl, A%dv + B%dv, A%du + B%du)` : Same as before, this will trigger unnecessary copies. Please revert this change.

Regarding the raising the error, please look at `linalg_state_type` and use it for consistency with the rest of `stdlib`.

File: src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp [hunk 3]
```diff
@@ -301,17 +302,21 @@ submodule (stdlib_specialmatrices) tridiagonal_matrices
...
     pure module function matrix_sub_tridiagonal_${s1}$(A, B) result(C)
         type(tridiagonal_${s1}$_type), intent(in) :: A
         type(tridiagonal_${s1}$_type), intent(in) :: B
         type(tridiagonal_${s1}$_type) :: C
-        C = tridiagonal(A%dl, A%dv, A%du)
-        C%dl = C%dl - B%dl; C%dv = C%dv - B%dv; C%du = C%du - B%du
+        
+        if (A%n /= B%n) error stop "ValueError: tridiagonal matrices must have the same dimension to be subtracted"
+        
+        C = tridiagonal(A%dl - B%dl, A%dv - B%dv, A%du - B%du)
```
[the remark this answers is not in the document]
Comment (loiseaujc) [reply src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp]: Same as before. Copilot is correct again.

**Review by loiseaujc — CHANGES_REQUESTED** [review]
Nice catch ! It never occurred to me that `1 x 1` matrices might be a particular edge case of tridiagonal matrices. It good overall, but there are a few bits and pieces that you need to revert to the original state for performance-sake, and look at `linalg_state_type` and `linalg_error_handling` to handle the errors is an `stdlib`-consistent way.

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 4]
```diff
@@ -146,8 +147,71 @@ contains
         #:endfor
     end subroutine
 
-end module
+    subroutine test_tridiagonal_1x1(error)
+        !> Test 1x1 matrix edge case for dense conversion
+        type(error_type), allocatable, intent(out) :: error
+        #:for k1, t1, s1 in (KINDS_TYPES)
+        block
```
Comment (jvdp1) [inline test/linalg/test_linalg_specialmatrices.fypp:377]: instead of templating the `block`s, I suggest to template the subroutine, and to not use `block`.

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 5]
```diff
@@ -146,8 +147,71 @@ contains
...
+            integer, parameter :: wp = ${k1}$
+            type(tridiagonal_${s1}$_type) :: A
+            ${t1}$, allocatable :: Amat(:,:), dl(:), dv(:), du(:)
+            
+            ! 1x1 matrix has 0-sized off-diagonals and 1-sized diagonal
+            allocate(dl(0), dv(1), du(0))
+            dv(1) = 5.0_wp
```
Comment (jvdp1) [inline test/linalg/test_linalg_specialmatrices.fypp]: `dl`, `dv` and `du` could be defined as `parameter`

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 6]
```diff
@@ -146,8 +147,71 @@ contains
...
+            integer, parameter :: n = 3
+            type(tridiagonal_${s1}$_type) :: A, B, C
+            ${t1}$, allocatable :: dl1(:), dv1(:), du1(:)
+            ${t1}$, allocatable :: dl2(:), dv2(:), du2(:)
+            
+            allocate(dl1(n-1), dv1(n), du1(n-1))
+            allocate(dl2(n-1), dv2(n), du2(n-1))
+            
+            dl1 = 1.0_wp ; dv1 = 2.0_wp ; du1 = 3.0_wp
+            dl2 = 4.0_wp ; dv2 = 5.0_wp ; du2 = 6.0_wp
```
Comment (jvdp1) [inline test/linalg/test_linalg_specialmatrices.fypp]: These variables could be defined as `parameter`

**Review by jvdp1 — COMMENTED** [review]
thank you for this PR and fix. Please find here some minor suggestions

File: test/linalg/test_linalg_specialmatrices.fypp (hunk 4 above)
Comment (loiseaujc) [reply test/linalg/test_linalg_specialmatrices.fypp:377]: I guess I'm partly responsible for this as, for unit test, I often tend to template the block and keep all of the same feature being tested (and irrespective of the kind or precision) in the same test routine. I don't really have a strong opinion about it though.

Comment (jvdp1) [reply test/linalg/test_linalg_specialmatrices.fypp:377]: I guess it is a matter of taste :) Using subroutines allows their parallelization, and therefore my preference. 
As the rest of the tests are coded with blocks, I suggest to keep them for homogenity.

**Review by jvdp1 — APPROVED** [review]
Thank you @srinjoy933  for this fix. I think that it is ready to be merged.

**Review by loiseaujc — APPROVED** [review]

File: src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp [hunk 7]
```diff
@@ -245,13 +243,18 @@ submodule (stdlib_specialmatrices) tridiagonal_matrices
...
-            B(i, i-1) = A%dl(i-1)
-            B(i, i) = A%dv(i)
-            B(i, i+1) = A%du(i)
-        enddo
-        B(n, n-1) = A%dl(n-1) ; B(n, n) = A%dv(n)
+        
+        if (n == 1) then
+            B(1, 1) = A%dv(1)
+        else
+            B(1, 1) = A%dv(1) ; B(1, 2) = A%du(1)
```
Comment (Mahmood-Sinan) [inline src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp]:
```suggestion
            B(1, 1) = A%dv(1)
            B(1, 2) = A%du(1)
```

File: src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp [hunk 8]
```diff
@@ -245,13 +243,18 @@ submodule (stdlib_specialmatrices) tridiagonal_matrices
...
+            do concurrent (i=2:n-1)
+                B(i, i-1) = A%dl(i-1)
+                B(i, i) = A%dv(i)
+                B(i, i+1) = A%du(i)
+            enddo
+            B(n, n-1) = A%dl(n-1) ; B(n, n) = A%dv(n)
```
Comment (Mahmood-Sinan) [inline src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp]:
```suggestion
            B(n, n-1) = A%dl(n-1)
            B(n, n) = A%dv(n)
```

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 9]
```diff
@@ -146,8 +152,90 @@ contains
         #:endfor
     end subroutine
 
-end module
+    subroutine test_tridiagonal_1x1(error)
+        !> Test 1x1 matrix edge case for dense conversion
+        type(error_type), allocatable, intent(out) :: error
+        #:for k1, t1, s1 in (KINDS_TYPES)
```
Comment (Mahmood-Sinan) [inline test/linalg/test_linalg_specialmatrices.fypp]:
```suggestion
        #:for k1, t1, s1 in KINDS_TYPES
```
Please fetch/pull once, because I've made some changes in the module. One of them was removing the brackets around KINDS_TYPES. For consistency, it would be better to remove the brackets.

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 10]
```diff
@@ -146,8 +152,90 @@ contains
...
+        block
+            integer, parameter :: wp = ${k1}$
+            type(tridiagonal_${s1}$_type) :: A
+            ${t1}$, allocatable :: Amat(:,:)
+            
+            ! Uses parameter with standard F2003 empty array syntax to satisfy maintainer
```
Comment (Mahmood-Sinan) [inline test/linalg/test_linalg_specialmatrices.fypp]: What do you mean by 'to satisfy maintainer' ?

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 11]
```diff
@@ -146,8 +152,90 @@ contains
...
+            call check(error, is_close(Amat(1,1), 5.0_wp), .true.)
+            if (allocated(error)) return
+        end block
+        #:endfor
+    end subroutine
 
+    subroutine test_tridiagonal_arithmetic(error)
+        !> Test arithmetic operations and optimization
+        type(error_type), allocatable, intent(out) :: error
+        #:for k1, t1, s1 in (KINDS_TYPES)
```
Comment (Mahmood-Sinan) [inline test/linalg/test_linalg_specialmatrices.fypp]:
```suggestion
        #:for k1, t1, s1 in KINDS_TYPES
```

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 12]
```diff
@@ -146,8 +152,90 @@ contains
...
+        block
+            integer, parameter :: wp = ${k1}$
+            integer, parameter :: n = 3
+            type(tridiagonal_${s1}$_type) :: A, B, C
+            
+            ! Uses explicit parameter lists to satisfy maintainer and avoid 'i' compiler bug
```
Comment (Mahmood-Sinan) [inline test/linalg/test_linalg_specialmatrices.fypp]: Again, 'to satisfy maintainer' ?

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 13]
```diff
@@ -146,8 +152,90 @@ contains
...
+            ${t1}$, parameter :: dl(0) = [${t1}$ ::]
+            ${t1}$, parameter :: du(0) = [${t1}$ ::]
+            ${t1}$, parameter :: dv(1) = [5.0_wp]
+            #:endif
+            
+            A = tridiagonal(dl, dv, du) 
+            Amat = dense(A)
+
+            ! Check if it compiled and converted properly without segfaulting
+            call check(error, size(Amat, 1) == 1, .true.)
```
Comment (Mahmood-Sinan) [inline test/linalg/test_linalg_specialmatrices.fypp]: How can compile error be detected by this code ? If there is a compile error, that would be detected while compiling/building, not while running.

**Review by Mahmood-Sinan — COMMENTED** [review]
@srinjoy933 Here are my views.

**Comment by srinjoy933** [comment]
hey @Mahmood-Sinan Thanks for your valuable suggestions. I have made the changes that you have suggested ,please have a look at this pr

File: test/linalg/test_linalg_specialmatrices.fypp (hunk 10 above)
Comment (srinjoy933) [reply test/linalg/test_linalg_specialmatrices.fypp]: Actually I have a habit to write comments to understand what changes I have made now , when I have already done multiple commits for a particular file.I forgot to erase off those comments in this case .Nothing else. Really thank you for pointing out this mistake .

File: test/linalg/test_linalg_specialmatrices.fypp (hunk 12 above)
Comment (srinjoy933) [reply test/linalg/test_linalg_specialmatrices.fypp]: same cause that I have previously mentioned

File: test/linalg/test_linalg_specialmatrices.fypp (hunk 13 above)
Comment (srinjoy933) [reply test/linalg/test_linalg_specialmatrices.fypp]: Thank you for pointing out this. You're completely right. I've adjusted the wording to simply state: ! Check if the 1x1 matrix converted properly at runtime without segfaulting

File: src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp [hunk 14]
```diff
@@ -36,7 +36,7 @@ submodule (stdlib_specialmatrices) tridiagonal_matrices
         !! Matrix dimension.
         type(tridiagonal_${s1}$_type) :: A
         !! Corresponding tridiagonal matrix.
-
+
```
Comment (Mahmood-Sinan) [inline src/specialmatrices/stdlib_specialmatrices_tridiagonal.fypp]: Please remove the trailing whitespace.

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 15]
```diff
@@ -113,7 +113,7 @@ contains
                 "spmv(fail): y = alpha*A*x + beta*y, alpha: "//to_string(alpha)//", beta: "//to_string(beta))
             if (allocated(error)) return
 
-            ! Test y = alpha * A.T @ x beta * y for random values of alpha and beta
```
Comment (Mahmood-Sinan) [inline test/linalg/test_linalg_specialmatrices.fypp:116]: Thanks for pointing it out.

File: test/linalg/test_linalg_specialmatrices.fypp [hunk 16]
```diff
@@ -383,7 +465,7 @@ program tester
     stat = 0
 
     testsuites = [ &
-        new_testsuite("specialmatrices", collect_suite) &
+        new_testsuite("special matrices", collect_suite) &
```
Comment (Mahmood-Sinan) [inline test/linalg/test_linalg_specialmatrices.fypp]:
```suggestion
        new_testsuite("special_matrices", collect_suite) &
```

**Review by Mahmood-Sinan — COMMENTED** [review]
@srinjoy933 Thanks for the rapid changes. A couple more changes and i think it will be good. One more suggestion is that,  since the 1x1 test and arithmetic test for tridiagonal uses parameter variables, would you mind doing the same for the 1x1 test and arithmetic test for sym tridiagonal ? It would be better for consistency.

**Comment by srinjoy933** [comment]
Thanks for the valuable suggestions, @Mahmood-Sinan! I'd really like to implement the same thing for symmetric tridiagonal matrices, but I think it would be best to handle that in a separate PR. If everything looks good to go here, we can merge this one and I'll tackle the symmetric tridiagonal updates next. Let me know what you think!

**Comment by srinjoy933** [comment]
@Mahmood-Sinan I have made the changes that you have suggested. Please have a look at this pr ,and kindly suggest if any more changes are required. Thank You!

**Review by Mahmood-Sinan — APPROVED** [review]
Thanks you for the PR. LGTM.

**Comment by jvdp1** [comment]
Thank you @srinjoy933 for this PR, and the reviewers as well, I will merge this PR.
